### PR TITLE
Improve messages for un-closed delimiter errors

### PR DESCRIPTION
**Pull request by estebank**

### CODE REVIEW

File: src/test/ui/parser/unclosed-braces.stderr [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  --> $DIR/unclosed-braces.rs:32:53
+   |
+LL | fn main() {
+   |           - un-closed delimiter
+...
+LL |         {
+   |         - this might be the culprit...
+...
+LL |     }
+   |     - ...as it matches this but it has different indentation
```
Comment (nikomatsakis) [inline src/test/ui/parser/unclosed-braces.stderr]: omg yes!

File: src/libsyntax/parse/lexer/tokentrees.rs [hunk 2]
```diff
@@ -49,9 +49,23 @@ impl<'a> StringReader<'a> {
...
+                    err.span_label(sp, "un-closed delimiter");
                 }
 
+                if let Some((delim, _)) = self.open_braces.last() {
+                    if let Some((d, open_sp, close_sp)) = self.suspicious_open_spans.iter()
+                        .filter(|(d, _, _)| delim == d)
+                        .next()  // these are in reverse order as they get inserted on close, but
+                    {            // we want the last open/first close
+                        if d == delim {
+                            err.span_label(*open_sp, "this might be the culprit...");
```
Comment (nikomatsakis) [inline src/libsyntax/parse/lexer/tokentrees.rs]: I love this language, but perhaps something a touch more formal...

- "this delimiter might not be properly closed" ?
- "this delimiter might be at fault" ?

**Review by nikomatsakis — APPROVED** [review]
I'm very excited about this. Looks like tests don't quite pass yet, though. 

I was wondering whether weirdness could happen because of macros, but I guess not: macro expansions can't introduce unpaired braces, after all. The error would occur when parsing the original file.

r=me once tests pass

**Comment by estebank** [comment]
> I'm very excited about this. Looks like tests don't quite pass yet, though.

It was a `parse-fail` test with outdated wording. Fixed.

> I was wondering whether weirdness could happen because of macros, but I guess not: macro expansions can't introduce unpaired braces, after all. The error would occur when parsing the original file.

The only thing that worries me is that code with wonky indentation will receive subpar suggestions, but I have banged my head trying to come up with better heuristics and they were all much more complicated than this for marginal benefit.

**Comment by nikomatsakis** [comment]
> The only thing that worries me is that code with wonky indentation will receive subpar suggestions...

I feel like that's ok.

**Comment by nnethercote** [comment]
[Benchmarking docs](https://github.com/rust-lang-nursery/rustc-perf/tree/master/collector).

**Pull request merged**